pyzfs version 0.2: support for extended API

avg-I opened the pull request:
Please **DO NOT MERGE**.
Note that this pull request is against pyzfs-0.1, not master.
This pull request is for reviewing only.
A new request should be created once pyzfs-0.1 is reviewed and merged.

Commit:
add support code for optional interfaces

Those interfaces are supported depending in whether the underlying
C library supports them.  That's intended for interfaces that are
still not committed and those that are not available for OpenZFS
platforms and actual versions.

Commit:
sphinx conf.py: add hack for having real signatures of decorated functions

The hack is for functools.wraps.

Commit:
lzc_promote: first uncommitted API extension

Commit:
lzc_promote: initial tests

Note that one of the tests currently fails because of an OpenZFS bug
(lack of validation).

Commit:
lzc extension: add bindings for the light-weight llibzfs_core extensions

Commit:
add wrappers based on light-weight lzc extensions api

These include:
- lzc_rename
- lzc_destroy
- lzc_inherit_prop
- lzc_set_prop
- lzc_get_props
- lzc_list_children
- lzc_list_snaps

Only generic error handling is added.
The tests are to follow.

Commit:
add a test case for extended lzc_receive functionality

The test case is disabled at the moment, it needs to be conditionalized
on whether underlying lzc_receive() can accept a bare dataset name
as a 'snapshot' parameter.

Commit:
bump version number to 0.2 for extended lzc API support

Commit:
switch to different libzfs_core extensions

Some notes:

_uncommitted decorator now takes an optional argument that makes
a status of a wrapped function dependent on another function.
This is needed to wrap several functions that all are implemented
via lzc_list.

nvlist_unpack() from libnvpair is exposed to the Python code, so
that we can unpack nvlists that we read from a file descriptor
returned by lzc_list().

Constants are defined for layout and size of the pipe record header.
Using Python struct module to unpack those records is much easier
than employing cffi for that.

It would probably make sense to turn lzc_list() into a context
manager, so that the pipe file descriptors (for both ends) could
be closed automatically.

Commit:
lzc_get_props: internally expand mountpoint property to its final value

At the moment mountpoint property is the only property for which its
default value can be determined in an algorithmic way.
Likewise, this is the only property where the further derivation of the
property value is required if its source is not the given dataset.

Commit:
lzc_get_props: highlight 'clones', 'mountpoint' and default properties

Commit:
catch up with ongoing libzfs_core api changes

Commit:
Merge branch 'pyzfs-0.2-lightweight-extensions' into pyzfs-0.2

Commit:
Merge branch 'pyzfs-0.1' into pyzfs-0.2-lightweight-extensions

Conflicts:
	libzfs_core/exceptions.py
	libzfs_core/_error_translation.py

Commit:
Merge branch 'pyzfs-0.2-lightweight-extensions' into pyzfs-0.2

Conflicts:
	libzfs_core/_error_translation.py
	libzfs_core/_libzfs_core.py

Commit:
fixup unqualified exception name

Commit:
catch up with another lizbsf_core API change

Commit:
recursion of depth 1 is required to list snapshots

Commit:
recursion of depth 1 is required to list a snapshot dataset

I am not sure if this is the expected behavior of
lzc_list() implementation.

Commit:
Merge branch 'pyzfs-0.1' into pyzfs-0.2-lightweight-extensions

Conflicts:
	libzfs_core/_error_translation.py

Commit:
catch up with nvlist_in() change

Commit:
Merge branch 'pyzfs-0.2-lightweight-extensions' into pyzfs-0.2

Conflicts:
	libzfs_core/_error_translation.py
	libzfs_core/_libzfs_core.py

Commit:
catch up with the nvlist_in() change

exarkun commented on docs/source/conf.py:
```diff
@@ -287,3 +287,18 @@
+        return func
+    return wrapper
+
+functools.wraps = no_op_wraps
```
Hmm.  This is probably rather unsafe.  It will only work as long as Sphinx itself doesn't depend on `functools.wraps` working properly - or if Sphinx manages to run all of its code that does depend on it before it evaluates this `conf.py`.

Both of these things might be true now but if that's the case this is still pretty fragile.  I'll try to make a follow-up suggestion for another way to address the problem this is aimed at (after I investigate a bit and see if there is another way).

exarkun commented on libzfs_core/_libzfs_core.py:
```diff
@@ -655,6 +660,276 @@ def lzc_exists(name):
     return bool(ret)
 
 
+def _uncommitted(check_func = None):
```
docstring for this and the resulting decorator function would definitely be good

exarkun commented on libzfs_core/_libzfs_core.py:
```diff
@@ -655,6 +660,276 @@ def lzc_exists(name):
+        fname = getattr(func, '__name__', None)
+        if fname is None:
+            raise ValueError('argument does not have __name__')
+        if not fname in globals():
```
`if x not in y:` reads more nicely

exarkun commented on libzfs_core/_libzfs_core.py:
```diff
@@ -655,6 +660,276 @@ def lzc_exists(name):
+            raise ValueError(fname + ' is not from libzfs_core')
+        if not callable(func):
+            raise ValueError(fname + ' is not a function')
+        func._is_supported = (getattr(_lib, fname, None) != None)
```
Usually you want `is not None` - http://stackoverflow.com/questions/11166748/identity-versus-equality-for-none-in-python gives a sort of good explanation about why (ignore the stuff about it "not having a value" - but the rest is valid)

exarkun replied:
What about `is_supported(is_supported)`?  It passes all these checks and reports `False` - which contradicts the documented behavior. :smile: 

Maybe a different check like `fname.startswith("lzc_")` or `fname in collection_of_functions_that_can_be_checked` would be a reasonable way to replace the globals/callable checks here.

avg-I replied:
For now, I am opting for checking both "lzc_" prefix and globals.

exarkun commented on libzfs_core/_libzfs_core.py:
```diff
@@ -655,6 +660,276 @@ def lzc_exists(name):
+    '''
+    try:
+        # woulf fail if is_supported is not set yet
+        return func._is_supported
```
I'm unsure the optimization of caching this result on the function object is worthwhile.  I suspect there won't be a lot of inner loops re-checking whether some function is supported.  And since the result won't ever change they should probably just hoist the check if they care about being faster (particularly since function calls in Python are often more expensive than the logic inside the function).

`is_supported` would be simpler and the `_is_supported` cache attribute would be avoided without the caching behavior, so I'd probably drop it.

exarkun commented on libzfs_core/_libzfs_core.py:
```diff
@@ -655,6 +660,276 @@ def lzc_exists(name):
+    '''
+    Destroy the ZFS dataset.
+
+    :param str name: the name of the dataset to destroy.
```
`bytes`, not `str` (as you did throughout the pyzfs-0.1 branch).

exarkun commented on libzfs_core/_libzfs_core.py:
```diff
@@ -655,6 +660,276 @@ def lzc_exists(name):
+
+
+@_uncommitted()
+def lzc_list(name, options):
```
A docstring would be good.

exarkun commented on libzfs_core/_libzfs_core.py:
```diff
@@ -655,6 +660,276 @@ def lzc_exists(name):
+    (rfd, wfd) = os.pipe()
+    fcntl.fcntl(rfd, fcntl.F_SETFD, fcntl.FD_CLOEXEC)
+    fcntl.fcntl(wfd, fcntl.F_SETFD, fcntl.FD_CLOEXEC)
+    options['fd'] = int32_t(wfd)
```
Mutating the dictionary passed in is usually not a good idea.  The caller might save or re-use the dictionary and be surprised at the change.

exarkun commented on libzfs_core/_libzfs_core.py:
```diff
@@ -655,6 +660,276 @@ def lzc_exists(name):
+
+
+@_uncommitted(lzc_list)
+def lzc_list_children(name):
```
I was surprised when `is_supported(lzc_list_children)` returned `False`.

avg-I replied:
I think that I have "outsmarted" myself with `is_supported` and `_uncommitted`... a fix is coming.

exarkun commented on libzfs_core/_libzfs_core.py:
```diff
@@ -655,6 +660,276 @@ def lzc_exists(name):
+
+
+@_uncommitted(lzc_list)
+def lzc_list_snaps(name):
```
Likewise for `is_supported(lzc_list_snaps)`.

exarkun commented on libzfs_core/_libzfs_core.py:
```diff
@@ -655,6 +660,276 @@ def lzc_exists(name):
+                child = result['name']
+                if child != name:
+                    yield child
+        finally:
```
`finally` inside a generator is a somewhat uncertain thing.  Either the generator must be iterated to exhaustion (presumably the caller) or the generator must be collected by the garbage collector.  If one of these things happens, this finally block will run.  If the generator ends up referenced from an uncollectable cycle, the finally block will never run.

It's very difficult to ensure objects don't end up in uncollectable cycles (particularly when writing code for someone else to use).  I tend to look at this feature of generators as being basically unusable.

One alternative would be to replace this generator with a normal function that collects results into a list and then returns the list.  This has the downside of greater memory use.  But it is at least straightforward.

The constant factor on memory usage could be reduced by just reading all the bytes and then using a generator to parse them and produce results.  The Python objects are probably much larger than the byte-stream representation of the same data so maybe this would even be a big win - not sure, probably easily measured.

Another option might be to make `lzc_list` return a ... context manager ... maybe?

``` py
with lzc_list(name, options) as listing:
    os.read(listing.read_fd, ...)
    ...
```

and have the two file descriptors closed by the context manager on exit.  Though if you go on to use this in a generator ...

``` py
def lzc_list_children(..):
    with lzc_list(name, options) as listing:
        stuff = os.read(listing.read_fd, ...)
        ...
        yield ...
```

Then actually you're probably back to the same old problem.  So nevermind that, I suppose.

Perhaps the sensible thing is to go with the memory-hungry solution and hope there's an optimization to be made later (either with or without an interface change...)

avg-I replied:
I never imagined that an iterator / generator could end up noncollectable.  There are, of course, many thing that I never imagined about Python before :smile:.
So, let me check if I have understood your suggestion (or rather, one of the options that you suggested) correctly: get the whole list from `lzc_list` in one go, then yield its elements one by one?

exarkun replied:
> get the whole list from lzc_list in one go, then yield its elements one by one?

Yep.  And, importantly, close the file descriptors before you start yielding.

avg-I committed:
Merge pull request #1 from ClusterHQ/pyzfs-0.1

pyzfs: a thin libzfs_core wrapper

Commit:
Merge branch 'pyzfs-0.1' into pyzfs-0.2-lightweight-extensions

Commit:
Merge branch 'pyzfs-0.2-lightweight-extensions' into pyzfs-0.2

Commit:
number of changes in is_supported()

- do not cache the result: this function should not be called in a hot path
  and the caller can always cache the result if needed
- reject functions that do not start with "lzc_", so that only libzfs_core
  C API funcitons are checked
- style fixes

Commit:
Merge branch 'pyzfs-0.2-lightweight-extensions' into pyzfs-0.2

Commit:
document _uncommitted() decorator

Commit:
move is_supported closer to related _uncommitted()

Commit:
Merge branch 'pyzfs-0.2-lightweight-extensions' into pyzfs-0.2

Conflicts:
	libzfs_core/_libzfs_core.py

Commit:
_uncommitted(): document check_func parameter

Commit:
fix is_supported() for functions that call differring libzfs_core functions

There are pyzfs API functions that are implemented on top of libzfs_core
API functions with a different name.
For example, lzc_list_children() and lzc_list_snaps() are both implemented
using lzc_list() from libzfs_core.

Previously, only _uncommitted() decorator was aware of such cases, but
is_supported() was not.  So, it would return False for lzc_list_snaps()
even when libzfs_core provided lzc_list().
Now, _uncommitted() decorator attaches the necessary information to
a decorated function and is_supported() checks that information to
evaluate the decorated function's availability.

Commit:
consistently use 'bytes' instead of 'str' for strings passed down to C

Commit:
Merge branch 'pyzfs-0.2-lightweight-extensions' into pyzfs-0.2

Commit:
document lzc_list

Commit:
lzc_list: do not mutate the options dictionary, use a copy

Commit:
zfs_list_{snaps,children}: clean up resources in a deterministic way

If the generator / iterator ends up being noncollectable, then the finally
section would never be executed and the resources won't get clean up.
Prevent that possibility by collecting all the data upfront and cleaning
up the resources before returning the iterator over the collected data.

The API is kept unchanged at the cost of the reduced efficiency.

Suggested by:	JP

Commit:
lzc_get_props: add a comment describing the work-around for ZFS-23

Commit:
add several sunny-day test cases

Commit:
Merge branch 'pyzfs-0.2-lightweight-extensions' into pyzfs-0.2

Commit:
fix up _uncommitted()

Commit:
catch up with an earlier rename

Commit:
change the work-around lzc_list() misbehavior when listing snapshots

We need to work around bug ZFS-26 for lzc_list() returning not only
snapshots but also immediate children when only snapshots are requested.

avg-I commented:
@exarkun I believe that I've addressed all of your comments in this PR.
Could you please double-check that?
I'd like to close this PR in favor of PR #6 (against master).

Commit:
lzc_receive: 'snapname' is allowed to be a filesystem name too

Commit:
Merge branch 'master' into pyzfs-0.2-lightweight-extensions

Commit:
Merge branch 'pyzfs-0.2-lightweight-extensions' into pyzfs-0.2

Commit:
improve documentation of new / uncommitted interfaces

Commit:
Merge branch 'pyzfs-0.2-lightweight-extensions' into pyzfs-0.2

Commit:
document another exception

Commit:
fix error translation for some new interfaces

Commit:
Merge branch 'pyzfs-0.2-lightweight-extensions' into pyzfs-0.2

Commit:
early error translation in lzc_list()

Commit:
subclass filesystem and snapshot exceptions from dataset ones

Specifically, inherit {Filesystem,Snapshot}{Exists,NotFound}
from Dataset{Exists,NotFound}.

Commit:
Merge branch 'master' into pyzfs-0.2-lightweight-extensions

Commit:
lzc_set_prop and lzc_inherit_prop refinements

Clarify that both can be applied to snapshots as well as to other dataset
types.
Clarify behavior for unknown and read-only properties.

Commit:
lzc_rename: document exception for nonexistent parent of the target

Commit:
add several test cases for the extended interfaces

Commit:
lzc_set_proper: clarify behavior for invalid property values

Document and test special behavior for too small quota values.

Commit:
use preferred python idiom instead of deprecated one

Commit:
clarify effect of running lzc_list_{children,snaps} on a snapshot

Commit:
lzc_get_props: correct error translation

This function can be called on all types of datasets including e.g.
snapshots.

Commit:
Merge branch 'pyzfs-0.2-lightweight-extensions' into pyzfs-0.2

exarkun commented:
Thanks.  Switching over to looking at #6 now.
